src/class/bth/bth_device.c

```diff
@@ -0,0 +1,238 @@
+    TU_ASSERT(desc_ep->bDescriptorType == TUSB_DESC_ENDPOINT);
+    dir = tu_edpt_dir(desc_ep->bEndpointAddress);
+    _btd_itf.ep_voice[dir] = desc_ep->bEndpointAddress;
+    // Store endpoint size for alternative 0
+    _btd_itf.ep_voice_size[dir][0] = (uint8_t) desc_ep->wMaxPacketSize.size;
+
+    for (int i = 1; i < CFG_TUD_BTH_ISO_ALT_COUNT; ++i) {
+      // Make sure rest of alternatives matches
+      itf_desc = (tusb_desc_interface_t const *)tu_desc_next(desc_ep);
+      TU_ASSERT(itf_desc->bDescriptorType == TUSB_DESC_INTERFACE);
+      TU_ASSERT(TUSB_CLASS_WIRELESS_CONTROLLER == itf_desc->bInterfaceClass &&
+                TUD_BT_APP_SUBCLASS == itf_desc->bInterfaceSubClass &&
+                TUD_BT_PROTOCOL_PRIMARY_CONTROLLER == itf_desc->bInterfaceProtocol);
+
+      desc_ep = (tusb_desc_endpoint_t const *)tu_desc_next(itf_desc);
+      dir = tu_edpt_dir(desc_ep->bEndpointAddress);
+      // Verify that alternative endpoint are same as first ones
+      TU_ASSERT(desc_ep->bDescriptorType == TUSB_DESC_ENDPOINT &&
+                _btd_itf.ep_voice[dir] == desc_ep->bEndpointAddress);
+      _btd_itf.ep_voice_size[dir][i] = (uint8_t) desc_ep->wMaxPacketSize.size;
+
+      desc_ep = (tusb_desc_endpoint_t const *)tu_desc_next(desc_ep);
+      dir = tu_edpt_dir(desc_ep->bEndpointAddress);
+      // Verify that alternative endpoint are same as first ones
+      TU_ASSERT(desc_ep->bDescriptorType == TUSB_DESC_ENDPOINT &&
+                _btd_itf.ep_voice[dir] == desc_ep->bEndpointAddress);
+      _btd_itf.ep_voice_size[dir][i] = (uint8_t) desc_ep->wMaxPacketSize.size;
+    }
+    // Consume all alternatives descriptors
+    (*p_length) += CFG_TUD_BTH_ISO_ALT_COUNT * (sizeof(tusb_desc_interface_t) + 2 * sizeof(tusb_desc_endpoint_t));
+  }
+
+  return true;
+}
+
+bool btd_control_complete(uint8_t rhport, tusb_control_request_t const *request)
+{
+  (void)rhport;
+
+  // Handle class request only
+  TU_VERIFY(request->bmRequestType_bit.type == TUSB_REQ_TYPE_CLASS);
+
+  if (tud_bt_hci_cmd_cb) tud_bt_hci_cmd_cb(&_btd_itf.hci_cmd, request->wLength);
+
+  return true;
+}
+
+bool btd_control_request(uint8_t rhport, tusb_control_request_t const *request)
+{
+  (void)rhport;
+
+  if (request->bmRequestType_bit.type == TUSB_REQ_TYPE_CLASS &&
+      request->bmRequestType_bit.recipient == TUSB_REQ_RCPT_DEVICE)
+  {
+    // HCI command packet addressing for single function Primary Controllers
+    TU_VERIFY(request->bRequest == 0 && request->wValue == 0 && request->wIndex == 0);
+  }
+  else if (request->bmRequestType_bit.recipient == TUSB_REQ_RCPT_INTERFACE)
+  {
+    if (request->bRequest == TUSB_REQ_SET_INTERFACE && _btd_itf.itf_num + 1 == request->wIndex)
+    {
+      // TODO: Set interface it would involve changing size of endpoint size
+    }
+    else // HCI command packet for Primary Controller function in a composite device
+        TU_VERIFY(request->bRequest == 0 && request->wValue == 0 && request->wIndex == _btd_itf.itf_num);
```

Comment:
we should open an `{ }` here, since TU_VERIFY() can screw up by expanding to multiple statements (although I have no intention to do that). We are still better to be safe.

Reply:
done